### PR TITLE
New kill fellow traitor objective!

### PR DESCRIPTION
**Pull request by beck-thompson**
## About the PR

Added a new traitor objective: kill fellow traitor. It should be a little less uncommon than kill target head objective (If someone has a better weight let me know. Its hard to choose 😆!)

## Why / Balance

This is part of my quest to make traitors not always buddy buddy! Now when a traitor comes to "help" you there should be just a *little* doubt in your mind that they are actually there to help. This shouldn't be a super common objective but just enough to make people worried.

You can still target a fellow traitor in the random person kill objective which is on purpose! Not knowing someone is a traitor and having to kill them is interesting.

Got idea from: #27872

## Media

image

**Changelog**

:cl:
- add: New traitor objective: kill fellow traitor!

### CODE REVIEW

**Comment by Dutch-VanDerLinde** [comment]
are traitors in the normal kill person pool? maybe remove them from that and replace it with this objective

**Comment by beck-thompson** [comment]
> are traitors in the normal kill person pool? maybe remove them from that and replace it with this objective

Yep they still are! As I said in the PR, I think it was interesting to still have them there as well. If people think otherwise though I'll remove em!

**Comment by DieselMohawk** [comment]
Is this just flavor text? If so I'm all for it.

**Comment by DieselMohawk** [comment]
Oh, I see, currently this makes kill objectives more common. Unsure how that will effect gameplay over the long run.

**Comment by beck-thompson** [comment]
> Oh, I see, currently this makes kill objectives more common. Unsure how that will effect gameplay over the long run.
Still cool though.

Yeah people don't like it being its own objective I'll just make it a flavor thing and make the text appear if the random kill objective turns out to be a traitor!

**Comment by DieselMohawk** [comment]
> > Oh, I see, currently this makes kill objectives more common. Unsure how that will effect gameplay over the long run.
> > Still cool though.
> 
> Yeah people don't like it being its own objective I'll just make it a flavor thing and make the text appear if the random kill objective turns out to be a traitor!

Alrighty, good luck on the PR

File: Resources/Prototypes/Objectives/traitor.yml [hunk 1]
```diff
@@ -111,6 +111,20 @@
     # if ce flies a shittle to centcom you better find a way onto it
     requireDead: true
 
+- type: entity
+  parent: [BaseTraitorObjective, BaseKillObjective]
+  id: KillRandomTraitorObjective
+  description: We have reason to beleve that they have become unloyal to the syndicate and need them elimninated.
```
Comment (deltanedas) [inline Resources/Prototypes/Objectives/traitor.yml:117]: believe, eliminated

**Review by deltanedas — CHANGES_REQUESTED** [review]

**Comment by deltanedas** [comment]
this doesnt make kill objectives more common, each objective group has its own weights

File: Resources/Prototypes/Objectives/traitor.yml (hunk 1 above)
Comment (beck-thompson) [reply Resources/Prototypes/Objectives/traitor.yml]: Yeah I'm bad at spelling thanks for the catch!

**Comment by DieselMohawk** [comment]
> this doesnt make kill objectives more common, each objective group has its own weights

Makes sense, thanks for the correction

**Comment by ThatOneGoblin25** [comment]
I feel like this is going to be rough on newer traitors who are automatically more trusting of fellow syndies.
It also feels like a slight nerf to antag activity since there's now an increased chance of needing to kill other syndie.
Otherwise nice pr I think it would be funny to roll midround traitor and have to kill a syndie who failed and is in perma.

**Comment by beck-thompson** [comment]
> I feel like this is going to be rough on newer traitors who are automatically more trusting of fellow syndies.
It also feels like a slight nerf to antag activity since there's now an increased chance of needing to kill other syndie.

Those are both valid points! However, I don't think they will effect the game *too* much. For the first point, I just don't think it will happen often enough to matter (Although if it does happen it would suck I fully agree). We also can get normal kill objectives on normal crew which I feel is a very similar issue if it targets a new player.

For the second point, that's kind of one of my big ideas of doing the PR! I personally have *never* seen syndicates attack each other and have only seen them team up. I'm hoping this PR will at least make them think a *little* bit before just teaming up willy nilly.

I DO NOT want this to stop traitors helping each other / teaming up though (Just make them think a little before doing so)! If you think it might be a little to common I'll decrease the chance of it happening so this isn't a big worry.

**Comment by whateverusername0** [comment]
you can already roll a kill objective for other traitors if you get lucky enough why make a separate one

**Comment by beck-thompson** [comment]
> you can already roll a kill objective for other traitors if you get lucky enough why make a separate one

That is very uncommon and also gives no indication that they are a traitor so 9/10 you will just treat it the same as a standard kill objective. This just makes it clear that they are a traitor and makes it slightly more frequent!

**Comment by Graded5** [comment]
Love this idea, I hate the instant trust given to other syndies. Only issue I see is syndies running to sec and telling them who the other syndie is, but I personally feel like thats not a huge issue.

**Comment by beck-thompson** [comment]
> Only issue I see is syndies running to sec and telling them who the other syndie is, but I personally feel like thats not a huge issue.

Yeah that's a fair concern but I also toggled the option so the target has to be dead, not just left on the station! Like you said, it seems risky for the other syndicate to reveal themselves as well so I don't think it would be super common / become an issue.

For a lot of this stuff, we kind of need to see how it plays out in game to really get a feeling for if its good or bad.

**Comment by ps3moira** [comment]
Traitors already barely work together and this feels like it'd kill almost any incentive to try to team up with syndicates

**Comment by beck-thompson** [comment]
> Traitors already barely work together and this feels like it'd kill almost any incentive to try to team up with syndicates

Where I play they work together somewhat frequently (And *never* betray / kill each other) so I'm seeing the opposite problem as you are. I also hope this addresses the basically free automatic (and kind of forced metagamey) trust someone gets when they approach you saying they have a help objective!

Honestly though, I really think we would need to see this in game to truly see if its helpful or harmful!

File: Resources/Prototypes/Objectives/traitor.yml [hunk 2]
```diff
@@ -111,6 +111,20 @@
     # if ce flies a shittle to centcom you better find a way onto it
     requireDead: true
 
+- type: entity
+  parent: [BaseTraitorObjective, BaseKillObjective]
+  id: KillRandomTraitorObjective
+  description: We have reason to believe that they have become unloyal to the syndicate and need them eliminated.
```
Comment (NakataRin) [inline Resources/Prototypes/Objectives/traitor.yml:117]: The word "unloyal" is not standard usage. While it is widely used, "disloyal" is the correct term. Up to you if you want to change it or not.

Comment (beck-thompson) [reply Resources/Prototypes/Objectives/traitor.yml]: I just decided to change the wording entirely! I didn't realize that was such an uncommon word.

**Comment by ElPidor** [comment]
questionable + its not really an argument that wher you play tots more communicative, so they should roll a kill a fellow traitor objective, sounds more of a secoff, SKILL issue. 
It would be good that second traitor who should be killed, should know that he have done something bad so killer is after him. Imagine head stuff, they dont trust some passenger because they metaknow that there's a tot and he's gonna kill his ass for objective.
Imagine 8 syndies on high pop and 3-4 of them can roll that objective.
Needs more discussion.
image

**Comment by Golinth** [comment]
I’d be okay with this if there was a chance for an opposite objective to spawn for the party being attacked to live and understand a syndie is hunting them. Basically worded as “You are being hunted by John Syndicate, Objective: Survive”

Otherwise as others have said, this basically just turns some syndies into valid sec officers

**Comment by ps3moira** [comment]
Pretty sure there's a upcoming PR https://github.com/space-wizards/space-station-14/pull/30198 that makes it so if you're in a certain organization, you're told to go kill that fellow traitor who is part of that organization. This PR wouldn't work if you're supposed to treat people in the same organization well but suddenly get a kill goal to kill them just for being a syndicate.

**Comment by beck-thompson** [comment]
> I’d be okay with this if there was a chance for an opposite objective to spawn for the party being attacked to live and understand a syndie is hunting them. Basically worded as “You are being hunted by fellow traitor John Syndicate, Objective: Survive”
> 
> Otherwise as others have said, this basically just turns some syndies into valid sec officers that may reduce the chaos in a round, instead of increase it.

I don't think I quite agree with you but I would be open as long as its not 100% of the time. I really am doing this so traitors don't always feel super safe to team up together! If they always knew they were the target of a kill mission then it would defeat the purpose entirely. However, it does seem like a lot of fun if they both know! It could lead to some crazy fights.

**Comment by beck-thompson** [comment]
> Pretty sure there's a upcoming PR https://github.com/space-wizards/space-station-14/pull/30198 that makes it so if you're in a certain organization, you're told to go kill that fellow traitor who is part of that enemy organization. This PR wouldn't work if you're supposed to treat people in the same organization well but suddenly get a kill goal to kill them just for being a syndicate.

Its a cool PR but I can't plan for stuff that isn't merged (And its still in draft to boot!). If it ever gets merged then I'll obviously update the logic but I can't do anything until that happens. I also think the objective would still totally work if they are in the same organization! If they are in the same organization the wording would remain basically the same:
`We have reason to believe that they have begun to question (the) [ORGANIZATION NAME] and need them eliminated.`

If they are a different organization, it could just be like "They are part of a rival syndicate group and must be eliminated." or something.

File: Content.Server/Objectives/Systems/KillPersonConditionSystem.cs [hunk 3]
```diff
@@ -99,6 +105,47 @@ private void OnHeadAssigned(EntityUid uid, PickRandomHeadComponent comp, ref Obj
         _target.SetTarget(uid, _random.Pick(allHeads), target);
     }
 
+    private void OnTraitorKillAssigned(EntityUid uid, PickRandomTraitorComponent comp, ref ObjectiveAssignedEvent args)
+    {
+        if (!TryComp<TargetObjectiveComponent>(uid, out var target))
+        {
+            Debug.Fail($"Missing components for {uid}.");
```
Comment (slarticodefast) [inline Content.Server/Objectives/Systems/KillPersonConditionSystem.cs:112]:
```suggestion
```
remove debugging code
Also, in all systems you can use Log.Debug("foobar") without having to import any additional namespaces.

File: Content.Server/Objectives/Systems/KillPersonConditionSystem.cs [hunk 4]
```diff
@@ -99,6 +105,47 @@ private void OnHeadAssigned(EntityUid uid, PickRandomHeadComponent comp, ref Obj
...
+            args.Cancelled = true;
+            return;
+        }
+
+        // target already assigned
+        if (target.Target != null)
+        {
+            Debug.Fail($"Target already assigned for {uid}.");
```
Comment (slarticodefast) [inline Content.Server/Objectives/Systems/KillPersonConditionSystem.cs:120]:
```suggestion
```
same here

File: Content.Server/Objectives/Systems/KillPersonConditionSystem.cs [hunk 5]
```diff
@@ -1,11 +1,14 @@
...
+using Content.Server.GameTicking.Rules;
 using Content.Server.Shuttles.Systems;
 using Content.Shared.CCVar;
 using Content.Shared.Mind;
 using Content.Shared.Objectives.Components;
 using Content.Shared.Roles.Jobs;
 using Robust.Shared.Configuration;
 using Robust.Shared.Random;
+using System.Diagnostics;
+using System.Linq;
```
Comment (slarticodefast) [inline Content.Server/Objectives/Systems/KillPersonConditionSystem.cs:11]: Not needed

File: Content.Server/Objectives/Systems/KillPersonConditionSystem.cs [hunk 6]
```diff
@@ -99,6 +105,47 @@ private void OnHeadAssigned(EntityUid uid, PickRandomHeadComponent comp, ref Obj
...
+        // No other traitors
+        if (traitors.Count == 0)
+        {
+            args.Cancelled = true;
+            return;
+        }
+
+        var targetTraitorMind = _random.Pick(traitors).Id;
+
+        // Make sure the current traitor doesn't have any targets targeting this traitor already.
```
Comment (slarticodefast) [inline Content.Server/Objectives/Systems/KillPersonConditionSystem.cs:136]: Wouldn't it make more sense to first filter out these cases and then pick from the remaining traitors list?

**Review by slarticodefast — CHANGES_REQUESTED** [review]
I still need to do some in-game testing, but it looks good from a coding perspective (except a few small things). I will bring it up for maintainer discussion.

**Pull request closed without merging**